// @sogni-ai/sogni-client/src/lib/RestClient.ts
import { ApiError, ApiErrorResponse } from '../ApiClient';
import TypedEventEmitter, { EventMap } from './TypedEventEmitter';
import { JSONValue } from '../types/json';
import { Logger } from './DefaultLogger';
import { AuthManager } from './AuthManager';

class RestClient<E extends EventMap = never> extends TypedEventEmitter<E> {
  readonly baseUrl: string;
  protected _auth: AuthManager;
  protected _logger: Logger;

  constructor(baseUrl: string, auth: AuthManager, logger: Logger) {
    super();
    this.baseUrl = baseUrl;
    this._auth = auth;
    this._logger = logger;
  }

  get auth(): AuthManager {
    return this._auth;
  }

  private formatUrl(relativeUrl: string, query: Record<string, string> = {}): string {
    const url = new URL(relativeUrl, this.baseUrl);
    Object.keys(query).forEach((key) => {
      url.searchParams.append(key, query[key]);
    });
    return url.toString();
  }

  private async request<T = JSONValue>(url: string, options: RequestInit = {}): Promise<T> {
    const init = await this.auth.authenticateRequest(options);

    // Add a timeout to detect hanging requests
    const controller = new AbortController();
    const timeoutId = setTimeout(() => {
      controller.abort();
    }, 30000);

    try {
      const response = await fetch(url, { ...init, signal: controller.signal });
      clearTimeout(timeoutId);
      return this.processResponse(response) as T;
    } catch (fetchError: any) {
      clearTimeout(timeoutId);
      throw fetchError;
    }
  }

  private async processResponse(response: Response): Promise<JSONValue> {
    let responseData;
    try {
      responseData = await response.json();
    } catch (e) {
      this._logger.error('Failed to parse response:', e);
      throw new Error('Failed to parse response');
    }
    // 401 means that the client instance is not authenticated, so we clear the authentication
    if (response.status === 401 && this.auth.isAuthenticated) {
      this.auth.clear();
    }
    if (!response.ok) {
      throw new ApiError(response.status, responseData as ApiErrorResponse);
    }
    return responseData as JSONValue;
  }

  get<T = JSONValue>(path: string, query: Record<string, any> = {}): Promise<T> {
    return this.request<T>(this.formatUrl(path, query));
  }

  post<T = JSONValue>(path: string, body: Record<string, unknown> = {}): Promise<T> {
    return this.request<T>(this.formatUrl(path), {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(body)
    });
  }
}

export default RestClient;
